### PR TITLE
Adapt bind_socket to make it usable with multiple processes

### LINKED ISSUE #428: UNIX domain socket raises 503 when workers are used
Is it Ok when a combination of `--workers ` and `--uds` raises 503 error via nginx? For example:
```
--workers 4 --uds /var/run/uvicorn-test.sock
```

it works well when `--workers` is omitted (1 worker). Please note that `gunicorn` works well via Unix socket and multiple workers.

Thanks!

### PR DESCRIPTION
**Pull request by euri10**
This makes it possible to use reload and workers flag with uds and fd.

fixes https://github.com/encode/uvicorn/issues/428 https://github.com/encode/uvicorn/issues/368 https://github.com/encode/uvicorn/issues/586 #722

### CODE REVIEW

**Comment by euri10** [comment]
still WIP I noticed the uvicorn logger message needs to be adapted

**Comment by euri10** [comment]
ready now @Kludex  !

File: uvicorn/config.py
Comment (Kludex) [inline uvicorn/config.py:454]: Does this impact the socket from `socket.fromfd`? https://docs.python.org/3/library/socket.html#socket.fromfd

image

File: uvicorn/config.py
Comment (Kludex) [inline uvicorn/config.py]: Those `logger.info` take a lot of lines now, can we have a single one at the end that accepts `logger_args` and the `color_message`?

This is just a small point not related to the logic, so you can close it if you think this is not necessary / overwork. :kissing:

File: tests/test_config.py
Comment (Kludex) [inline tests/test_config.py:372]: Is there a reason for not adding any test for `fd`?

**Review by Kludex — COMMENTED** [review]
I did my best here. It's a short PR, but I feel like I'm not qualified to approve it yet. I'm going to study some details before having the confidence to do so.

But I hope my current comments make sense.

File: uvicorn/config.py
Comment (euri10) [reply uvicorn/config.py:454]: this `sock.set_inheritable(True)` applies to all 3 cases so yes. it serves when we have multiple process and 1 socket to listen on iirc

File: uvicorn/config.py
Comment (euri10) [reply uvicorn/config.py]: yes that's a good idea, done

File: tests/test_config.py
Comment (euri10) [reply tests/test_config.py:372]: no, I added one

**Comment by euri10** [comment]
ready for another review @Kludex 
let me know if something not clear I can easily expain

File: uvicorn/config.py
Comment (Kludex) [reply uvicorn/config.py:454]: Ah, I understand what confused me now. Thanks.

File: uvicorn/main.py
Comment (Kludex) [inline uvicorn/main.py]: Maybe I'm mistaken, but can't we have a single one of those outside the conditional?

```suggestion
    if config.uds:
        os.remove(config.uds)
```

Comment (Kludex) [reply uvicorn/main.py]: I guess we can, but it's not right, because we bind inside them. :thinking: Right?

Comment (euri10) [reply uvicorn/main.py]: it looks simpler yes, I removed occurrences in the basereload and multiprocess shutdown method and put it outside

**Review by Kludex — APPROVED** [review]
I'm confident enough now.

But maybe we should wait for the `mypy` merge first? Anyway, your call.

**Comment by euri10** [comment]
> But maybe we should wait for the `mypy` merge first? Anyway, your call.

the one on config.py ?

**Comment by Kludex** [comment]
Yeah

**Comment by euri10** [comment]
> Yeah

that's a good idea, he put a lot of efforts into it and it would be cruel to have to add that part on top of it ;)

File: uvicorn/config.py
Comment (euri10) [inline uvicorn/config.py:398]: just needed to add this post-merge with master @Kludex 
the rest stayed the same

Comment (euri10) [reply uvicorn/config.py:398]: So  I removed the hold button and feel free to merge if still ok with this !

Comment (Kludex) [reply uvicorn/config.py:398]: I find it so weird that the approval is not removed after a commit is pushed. I don't understand the reason behind it.

Anyway, LGTM! :)

**Pull request merged**
